socorro/external/boto/connection_context.py

```diff
@@ -0,0 +1,245 @@
+from configman.converters import class_converter, py_obj_to_str
+from socorro.lib.util import DotDict
+from socorro.lib.converters import change_default
+
+class BotoNotFound(Exception):
+    pass
+
+
+#==============================================================================
+class ConnectionContext(RequiredConfig):
+
+    required_config = Namespace()
+    required_config.add_option(
+        'host',
+        doc="The hostname (leave empty for AWS)",
+        default="",
+        reference_value_from='resource.boto',
+    )
+    required_config.add_option(
+        'port',
+        doc="The network port (leave at 0 for AWS)",
+        default=0,
+        reference_value_from='resource.boto',
+    )
+    required_config.add_option(
+        'access_key',
+        doc="access key",
+        default="",
+        reference_value_from='resource.boto',
+    )
+    required_config.add_option(
+        'secret_access_key',
+        doc="secret access key",
+        default="",
+        secret=True,
+        reference_value_from='secrets.boto',
+    )
+    required_config.add_option(
+        'bucket_name',
+        doc="The name of the bucket.",
+        default='crashstats',
+        reference_value_from='resource.boto',
+        likely_to_be_changed=True,
+    )
+    required_config.add_option(
+        'prefix',
+        doc="a prefix to use inside the bucket",
+        default='',
+        reference_value_from='resource.boto',
+        likely_to_be_changed=True,
+    )
+    required_config.add_option(
+        'calling_format',
+        doc="fully qualified python path to the boto calling format function",
+        default='boto.s3.connection.SubdomainCallingFormat',
+        from_string_converter=class_converter,
+        reference_value_from='resource.boto',
+        likely_to_be_changed=True,
+    )
+
+    operational_exceptions = (
+        socket.timeout,
+        # wild guesses at retriable exceptions
+        boto.exception.PleaseRetryException,
+        boto.exception.ResumableTransferDisposition,
+        boto.exception.ResumableUploadException,
+    )
+
+    conditional_exceptions = (
+        boto.exception.StorageResponseError
+    )
+
+    #--------------------------------------------------------------------------
+    def is_operational_exception(self, x):
+        if "not found, no value returned" in str(x):
+            # the not found error needs to be re-tryable to compensate for
+            # eventual consistency.  However, a method capable of raising this
+            # exception should never be used with a transaction executor that
+            # has infinite back off.
+            return True
+        #elif   # for further cases...
+        return False
+
+
+    #--------------------------------------------------------------------------
+    def __init__(self, config, quit_check_callback=None):
+        self.config =  config
+
+        self._connect_to_endpoint = boto.connect_s3
```

Comment:
The class is all about `boto`. Not about `S3` yet this is hardcoded as the default.